### PR TITLE
Ruby - Tic-Tac-Toe - Game Logic/Milestone 3

### PR DESCRIPTION
**Pull request by Rhelli**
## Pull request for review by Microverse TSE's. 
This newly created **game_logic** branch will be merged into the current working *development* branch. Included in this branch is:

> Player classes allowing initialisation/creation of players

> A cell class storing values for 'X' or 'O'

> A Board class to:
        -  Initialise/create a new, empty board each time a game starts
        -  Fetch and place player pieces on the board
        -  Define if a game has ended in either a win or a draw

> A Game class to:
        -  Initialise/create a new game and direct the players through their turns
        -  Alternation of players after each turn
        -  Randomisation of the starting player
        -  Player generator (with 'gets' name)
        -  Mapping of human input (1-9) into coordinates
        -  A Game over method
        -  A visual tic-tact-toe board that the place can see and place their pieces on
        -  Optional instruction display
        -  A rematch option to restart the game

Additionally, there are various arrays of responses to the user, randomised, giving the user a different experience every time they play.

### CODE REVIEW

File: bin/main [hunk 1]
```diff
@@ -1,107 +1,52 @@
 #!/usr/bin/env ruby
 
-#Game Startup
+# frozen_string_literal: true
+
+require './lib/game_logic.rb'
+
```
Comment (imhta) [inline bin/main]: Fix this issue too
```bash
$ ruby ./main 
Traceback (most recent call last):
        2: from ./main:5:in `<main>'
        1: from C:/Ruby26-x64/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
C:/Ruby26-x64/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require': cannot load such file -- ./lib/game_logic.rb (LoadError)
```

**Review by imhta — CHANGES_REQUESTED** [review]
# Changes required 🛑 

### Hello  @Rhelli @queeksm  , 👋 You have done a great job in this project UI 👏 I loved ❤️ this project interface. But still few things need to be done before moving to next milestone.

- [ ] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
 
- kindly check the inline comment too 

----
### Don't forgot to request for another code review
----
**Pon Muthu Selvam N**
__nponmuthuselvam@gmail.com__
_If you have any queries, feel free to ping me on slack_

**Comment by shshamim63** [comment]
## Status :recycle: [!! Changes Required !!] :recycle: 
___
Hi @Rhelli and @queeksm :wave: ,
Well I can see that you haven't covered the issues mentioned by the previous reviewer @imhta .
So please follow the instructions below :point_down: and if you have any different idea on it, don't forget to share it. 
___
## :spades: [Mandatory Changes] :spades: 
___
1. Please delete all the commented code.
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/bin/main#L50-L62
2. All the UI related code should take place in main.rb file. but I can see puts on your game_logic.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/lib/game_logic.rb#L108-L109 
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/lib/game_logic.rb#L120-L122
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/lib/game_logic.rb#L151-L160
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/lib/game_logic.rb#L190-L365
3. I went through the same issue that has been mentioned above by @imhta realated to `require` . 
 `require_relative` should be used for referring to files within your directory
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/bin/main#L5
4. When you want to disable a rubocop linting rules(**It is better not to do it**) . Use rubocop.yml to disable them.
https://github.com/queeksm/Tic-Tac-Toe/blob/a3f2ec98ee5f95e3e3f4e9540bc68f296101c8b0/lib/game_logic.rb#L3-L21
___
#### Please fix the above mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Slack](__Shakhawat Hossain__)

File: lib/game_logic.rb [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+  def ask_move_error
+    [
+      "#{@player_one.name}, please enter a number from 1 to 9 or select a cell that it's not taken. This is Tic-Tac-toe nort rocket science...",
+      "#{@player_one.name}, please select a valid number between 1 and 9 that #{@player_two.name} hasn't already chosen.",
+      "#{@player_one.name}, why are you doing this to me? Don't you know that you can only choose a number between 1 and 9?",
+      "#{@player_one.name}, why can't you be nice like #{@player_two.name} and enter a number between 1 and 9 that isn't already taken?!",
+      "#{@player_one.name}, are you taunting me? I may be a machine but please pick a number between 1 and 9 that #{@player_two.name} has not already chosen...Please.."
+    ]
+  end
```
Comment (whiz25) [inline lib/game_logic.rb:98]: - this too. any user interface code should not be in this file.
- No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: lib/game_logic.rb [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+      ██╗   ██╗ ██████╗ ██╗   ██╗    ██████╗  ██████╗ ████████╗██╗  ██╗    ███████╗██╗   ██╗ ██████╗██╗  ██╗██╗
+      ╚██╗ ██╔╝██╔═══██╗██║   ██║    ██╔══██╗██╔═══██╗╚══██╔══╝██║  ██║    ██╔════╝██║   ██║██╔════╝██║ ██╔╝██║
+       ╚████╔╝ ██║   ██║██║   ██║    ██████╔╝██║   ██║   ██║   ███████║    ███████╗██║   ██║██║     █████╔╝ ██║
+        ╚██╔╝  ██║   ██║██║   ██║    ██╔══██╗██║   ██║   ██║   ██╔══██║    ╚════██║██║   ██║██║     ██╔═██╗ ╚═╝
+         ██║   ╚██████╔╝╚██████╔╝    ██████╔╝╚██████╔╝   ██║   ██║  ██║    ███████║╚██████╔╝╚██████╗██║  ██╗██╗
+         ╚═╝    ╚═════╝  ╚═════╝     ╚═════╝  ╚═════╝    ╚═╝   ╚═╝  ╚═╝    ╚══════╝ ╚═════╝  ╚═════╝╚═╝  ╚═╝╚═╝
+
+         ")
+    end
+  end
```
Comment (whiz25) [inline lib/game_logic.rb:169]: - move this also to the main.rb file
- No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: lib/game_logic.rb [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        'Thank you, come again!',
+        '[This is the end music intensifies]',
+        "Thanks for playing Andres and Rory's Tic-Tac-Toe!",
+        'See you soon!',
+        'Thanks for playing!',
+        'We hope you enjoyed the game! Goodbye!'
+      ]
+      outputs.sampler(rematch_no_responses)
+    end
+  end
```
Comment (whiz25) [inline lib/game_logic.rb:278]: - any code to do with user interface should be in the main.rb not here

File: lib/game_logic.rb [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+      "#{@player_one.name}, please enter a number from 1 to 9 to choose your position.",
+      "#{@player_one.name}, please select your next spot with a number between 1 and 9.",
+      "#{@player_one.name}, its you're turn! Choose a number between 1 and 9 to place your symbol.",
+      "#{@player_one.name}, time to pick a spot! Select a number 1 to 9 to choose your next move.",
+      "#{@player_one.name}, pick dat dere spot on dat dere board, wit dem numbers 'tween 1 and 9.",
+      "#{@player_one.name}, there are 9 spots. Many of these spots have been taken, the world waits, for you, at this very moment,
+      to make the ultimate decision. The ultimate sacrifice. Right here. Right now.\n\n Please enter a number from 1 to 9 to choose your position.",
+      "#{@player_one.name}, life is like a box of chocolates. Please enter a number from 1 to 9 to choose your position."
+    ]
+  end
```
Comment (whiz25) [inline lib/game_logic.rb:88]: - No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)). So move that part of code to the `main.rb`

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job guys :+1: 
You have done well. However, we need to do a few things before we can move on.
Check my comments and implement those changes.

### Martin Afani

**Comment by Rhelli** [comment]
@whiz25 @imhta @shshamim63

Hi there, thanks for the code review guys - we've just got a couple of things that we're a little confused about as from our point of view as we believe we have fulfilled the 'Microverse Milestone 3: Game Logic' requirements. I believe the main problem is over the statement below 👇🏽

Screenshot 2020-01-07 at 11 05 09

What appears clear to us from the above is that any user interaction which uses puts, print or gets.chomp should not be included in the game_logic.

Our previous PR unfortunately _**did**_ include some print and puts in the game logic but, since that version and in order to not have to completely rewrite a fully functioning version of the game, we created methods for **puts (putter), print (printer) and get.chomp (chomper)** in order to fix that. Thus, from our point of view there aren't actually any instances of puts, print of get.chomp in the game_logic but simply calls to methods in the main.rb which do use puts, print and get.chomp.

The last thing is over the response arrays (the arrays which include multiple responses for the player) - these similarly use these putter, printer and chomper methods created in main.rb.

Let us know your thoughts.

**Review by uche-inyama — APPROVED** [review]
## Approved

Hi team!

I have read your feedback on the comment; I believe you have taken correction.

I will go ahead and approve this project.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Comment by Rhelli** [comment]
@uche-inyama 

Amazing 🤩 Thanks for your time. Have a good day!

Rory

**Pull request merged**